src/keymap.rs

```diff
@@ -13,6 +14,58 @@ use crate::tty::{RawReader, Term, Terminal};
 /// The number of times one command should be repeated.
 pub type RepeatCount = usize;
 
+/// Custom dynamic action
+#[derive(Clone)]
+pub struct Action {
+    pub name: String,
+    /// Takes the currently edited `line` with the cursor `pos`ition and
+    /// returns the command to be performed or `None` to perform the default
+    /// one.
+    pub action: fn(line: &str, pos: usize, ctx: &ActionContext) -> Option<Cmd>,
```

Comment:
I feel like the `line` and `pos` should be part of the context.

An `fn` is pretty restrictive because it doesn't allow adding any of your own context. Having something like
```
Arc<Box<dyn Fn(&ActionContext) -> Option<Cmd>>>
```
would be much more flexible.

Also I think it would be better to not make this public to allow non-breaking changes in the future.

Reply:
Most of the time, you just need to check `line` and/or `pos`. So if you can access them directly, it seems better/easier.
And could you provide a case where an action is not stateless ?

Reply:
For example I've been playing around with having a key bindings to insert bits of text with [`skim`](https://github.com/lotabout/skim). The bits of text can be dependent on the rest of the program.

Reply:
We have just merged `Validation` where anyone would use `line` for sure, but it only has `ValidationContext` in the signature. Well, the reason for validator is probably because line is mutable there. But I think having similar signature is nice anyway.